### PR TITLE
[jvm-packages][xgboost4j-gpu] Add option to load data from GPU memory

### PR DESCRIPTION
**Pull request by wbo4958**
This PR adds GPU support for XGBoost4j-gpu by leveraging the latest CUDF jar release.

With this PR, Users can do the ETL for the local CSV/Parquet/Orc files by CUDF jar, then build
the GpuDMatrix and do the training.

All of the above stages are accelerating by GPUs.

### CODE REVIEW

**Comment by wbo4958** [comment]
Hi @trivialfis @hcho3, The gpu test case failed on docker env. I locally repro it  with the Docker with the below error

``` console
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  for_each: failed to synchronize: cudaErrorInvalidDevice: invalid device ordinal
```

Any ideas on this failure.

BTW, I can pass all test cases without Docker locally.

**Comment by wbo4958** [comment]
@hcho3 @trivialfis Could you help to review this PR?

File: jvm-packages/CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +23,16 @@ target_include_directories(xgboost4j
   ${PROJECT_SOURCE_DIR}/dmlc-core/include
   ${PROJECT_SOURCE_DIR}/rabit/include)
 
+if (USE_CUDA)
```
Comment (trivialfis) [inline jvm-packages/CMakeLists.txt:26]: This is unnecessary as it's handled in top level CMake file.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.util.Iterator;
+
+import ml.dmlc.xgboost4j.java.DMatrix;
+import ml.dmlc.xgboost4j.java.XGBoostError;
+
+/**
+ * DMatrix based on GPU
+ */
+class ColumnDMatrix extends DMatrix {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:27]: Why is it called colum matrix?  The comment in constructor uses `DeviceQuantileDMatrix`.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  /**
+   * Create the normal DMatrix from column array interface
+   * @param featureArrayInterfaces the cuda array interface of feature columns
+   * @param missing missing value
+   * @param nthread threads number
+   * @throws XGBoostError
+   */
+  public ColumnDMatrix(String featureArrayInterfaces, float missing, int nthread)
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:54]: Do we need to expose this?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    handle = out[0];
+  }
+
+  /**
+   * Set label of DMatrix from cuda array interface
+   *
+   * @param labelJson the cuda array interface of label column
+   * @throws XGBoostError native error
+   */
+  public void setLabel(String labelJson) throws XGBoostError {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:69]: Can this be implemented in DMatrix instead?  We can just throw a runtime error if CUDA is not enabled.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        XGBoostJNI.XGDMatrixSetInfoFromInterface(handle, "label", labelJson));
+  }
+
+  /**
+   * Set weight of DMatrix from cuda array interface
+   *
+   * @param weightJson the cuda array interface of weight column
+   * @throws XGBoostError native error
+   */
+  public void setWeight(String weightJson) throws XGBoostError {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:80]: Ditto.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        XGBoostJNI.XGDMatrixSetInfoFromInterface(handle, "weight", weightJson));
+  }
+
+  /**
+   * Set base margin of DMatrix from cuda array interface
+   *
+   * @param baseMarginJson the cuda array interface of base margin column
+   * @throws XGBoostError native error
+   */
+  public void setBaseMargin(String baseMarginJson) throws XGBoostError {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:91]: Ditto.  Also, why are we using strings directly?

File: jvm-packages/xgboost4j-gpu/src/test/resources/iris.data.csv [hunk 7]
```diff
@@ -0,0 +1,150 @@
+5.1,3.5,1.4,0.2,0.0
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/test/resources/iris.data.csv:1]: Please fetch any test data you need instead of committing them into git history.

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/NativeLibLoader.java [hunk 8]
```diff
@@ -118,7 +118,7 @@ static Arch detectArch() {
    * Throws UnsatisfiedLinkError if the library failed to load it's dependencies.
    * @throws IOException If the library could not be extracted from the jar.
    */
-  static synchronized void initXGBoost() throws IOException {
+  public static synchronized void initXGBoost() throws IOException {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/NativeLibLoader.java:121]: Why is this necessary?

File: jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h [hunk 9]
```diff
@@ -0,0 +1,17 @@
+#ifndef JVM_UTILS_H_
+#define JVM_UTILS_H_
+
+#include "./xgboost4j-gpu.h"
+
+#define JVM_CHECK_CALL(__expr)                                                 \
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h:6]: I think we have an utility inside the original jvm wrapper header right?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/XGBoostJNI.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      throw new XGBoostError(XGBGpuGetLastError());
+    }
+  }
+
+  private final static native String XGBGpuGetLastError();
+
+  public final static native int XGDMatrixSetInfoFromInterface(
+    long handle, String field, String json);
+
+  public final static native int XGDeviceQuantileDMatrixCreateFromCallback(
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/XGBoostJNI.java:53]: Can this be part of the original JNI.java?  Just a question.

**Review by trivialfis — COMMENTED** [review]
Preliminary review comments.  I'm not an expert at JVM so most of them are just questions instead of change requests.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 2 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:27]: DMatrix created by columnars. Any suggestions?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 3 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:54]: Yeah, most of the real cases may not include base_margin or weight.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 4 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:69]: Better to keep this in Gpu side. Is there any plan to support this in CPU ?

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      float[] anchor = convertFloatTofloat(labelFloats);
+      float[] label = dMatrix.getLabel();
+      float[] weight = dMatrix.getWeight();
+      float[] baseMargin = dMatrix.getBaseMargin();
+
+      TestCase.assertTrue(Arrays.equals(anchor, label));
+      TestCase.assertTrue(Arrays.equals(anchor, weight));
+      TestCase.assertTrue(Arrays.equals(anchor, baseMargin));
+    } catch (XGBoostError xgBoostError) {
+      xgBoostError.printStackTrace();
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java:50]: yeah, getLabel/getWeights ... will throw XGBoostError.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/XGBoostJNI.java (hunk 10 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/XGBoostJNI.java:53]: Yes, we can. But we need to involved some class in XGBoost4j package.

File: jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h (hunk 9 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h:6]: yeah, just this one.

File: jvm-packages/xgboost4j-gpu/src/native/xgboost4j-gpu.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
+#include <xgboost/c_api.h>
+#include "./xgboost4j-gpu.h"
+#include "jvm_utils.h"
+
+namespace xgboost {
+namespace jni {
+    jint XGDeviceQuantileDMatrixCreateFromCallbackImpl(JNIEnv *jenv, jclass jcls,
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/native/xgboost4j-gpu.cpp:7]: The implementation is in xgboost::jni namespace

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    //set params
+    Map<String, Object> paramMap = new HashMap<String, Object>() {
+      {
+        put("max_depth", 2);
+        put("objective", "multi:softprob");
+        put("num_class", 3);
+        put("num_round", round);
+        put("num_workers", 1);
+        put("tree_method", "gpu_hist");
+        put("predictor", "gpu_predictor");
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:64]: No big deal, right?

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java:50]: If it's caught then the test won't fail, right?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 2 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:27]: The underlying implementation is not limited to columnar matrix, it can be arbitrary data type.  Let's not restrict ourselves to cudf.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 3 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:54]: I meant do we need to expose an interface that accepts string instead of real data?

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 4 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:69]: Not yet.

File: jvm-packages/xgboost4j-gpu/src/native/xgboost4j-gpu.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/native/xgboost4j-gpu.cpp:7]: Can we move it out?

File: jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h (hunk 9 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/native/jvm_utils.h:6]: I meant we have the same code in xgboost4j/src/native right?

File: jvm-packages/CMakeLists.txt (hunk 1 above)
Comment (wbo4958) [reply jvm-packages/CMakeLists.txt:26]: Done

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 4 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:69]: So, let's just keep the interface here. I don't want to touch CPU code if unnecessary.

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java:50]: you're right.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 3 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:54]: Yeah, better to expose this. then users can use this API to construct DMatrix from array interface

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 2 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:27]: any suggestion for the new name? Note: This class may create the Normal DMatrix or DeviceDMatrix

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrixTest.java]: Done

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 3 above)
Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java:54]: I meant, why not

``` java
public ColumnMatrix(Table data, float missing) {
    String str = getArrayInterface(data)
    ...
}
```

Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java]: done

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java (hunk 5 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/ColumnDMatrix.java]: done

**Comment by trivialfis** [comment]
Restarted the CI.

**Comment by wbo4958** [comment]
@hcho3 @trivialfis  any further suggestions?

**Comment by trivialfis** [comment]
Will get back to this next week.  Thanks for the patience.

**Comment by trivialfis** [comment]
https://github.com/wbo4958/xgboost/pull/3

**Comment by trivialfis** [comment]
https://github.com/wbo4958/xgboost/pull/4

**Comment by trivialfis** [comment]
Will push some changes tomorrow.

**Review by RAMitchell — CHANGES_REQUESTED** [review]
Thanks @wbo4958.

I think the DMatrix constructor needs reworking. Putting the same cudfdataframe object in for both X and y is strange.

**Comment by wbo4958** [comment]
> Thanks @wbo4958.
> 
> I think the DMatrix constructor needs reworking. Putting the same cudfdataframe object in for both X and y is strange.

Done

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java [hunk 14]
```diff
@@ -83,18 +83,19 @@ public void testBooster() throws XGBoostError {
 
     try (Table tmpTable = Table.readCSV(schema, opts,
 					new File(trainingDataPath))) {
-      CudfDataFrame cudfDataFrame = new CudfDataFrame(tmpTable, new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}, new int[]{12});
+      CudfColumnBatch batch = new CudfColumnBatch(tmpTable, new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}, new int[]{12});
+      CudfColumn labelColumn = CudfColumn.from(tmpTable.getColumn(12));
```
Comment (RAMitchell) [inline jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:87]: > , new int[]{12}

These label indices should not be here right?

**Review by RAMitchell — COMMENTED** [review]
Changes look good. I will leave it to others to give final approval.

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+  /**
+   * CudfColumnBatch constructor
+   * @param table             the CUDF table
+   * @param featureIndices    must-have, specify the feature's indices in the table
+   * @param labelIndices      must-have, specify the label's indices in the table
+   * @param weightIndices     optional, specify the weight's indices in the table
+   * @param baseMarginIndices optional, specify the base marge's indices in the table
+   */
+  public CudfColumnBatch(Table table, int[] featureIndices, int[] labelIndices, int[] weightIndices,
```
Comment (RAMitchell) [inline jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: I think this class should only contain feature columns, and the additional columns should be added like

> CudfColumn labelColumn = CudfColumn.from(tmpTable.getColumn(12));

Or is there performance reason why it was done like this?

Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: This class will be used to create DeviceDMatrix which requires feature/label (must-have) and weight/base_margin (optional)

File: jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java (hunk 14 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/test/java/ml/dmlc/xgboost4j/gpu/java/BoosterTest.java:87]: the batch will be used to create the Normal DMatrix which does not require label column and the DeviceDmatrix which require label column

File: jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java (hunk 15 above)
Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: CudfColumnBatch will be used to construct Normal Dmatrix and DeviceDMatrix. For that purpose, I define 3 different constructors for it. the first one is only accepting features which will be used to create normal dmatrix. and the others are for creating DeviceDMatrix. So it doesn't matter here.

Comment (RAMitchell) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Can you have a different constructor for DeviceDMatrix requiring these inputs upfront?

Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Hi @RAMitchell, what does that mean?

Comment (RAMitchell) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Here is what I imagine it could look like:
```
CudfColumnBatch batch = new CudfColumnBatch(tmpTable, new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11});
CudfColumn labelColumn = CudfColumn.from(tmpTable.getColumn(12));

DMatrix dMatrix1 = new DMatrix(batch, Float.NaN, 1);
dMatrix1.setLabel(labelColumn)

// Require this class to include meta info vectors in constructor
DeviceDMatrix dMatrix2 = new DeviceDMatrix(batch, Float.NaN, 1, labelColumn, null, null);
```
Also, if you can do the slicing of tmpTable using native cudf functionality it's better.

e.g.
```
// CudfColumnBatch just accepts a dataframe and we use cudf to separate the parts we want
CudfColumnBatch batch = new CudfColumnBatch(tmpTable.getColumns(new int[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}));
```
I'm assuming here we can do this without actually copying any data, let me know if that assumption is incorrect.

Comment (wbo4958) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Is there any necessary to separate DeviceDmatrix from DMatrix? 

And BTW, there is not involving any copying data from GPU to CPU ...

Comment (RAMitchell) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Remind me why does the DeviceDmatrix need the label column in the constructor and DMatrix doesn't?

Comment (trivialfis) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: @RAMitchell I think `DeviceQuantileDMatrix` can only be used for training while `DMatrix` can be used for other puroses.

Comment (RAMitchell) [reply jvm-packages/xgboost4j-gpu/src/main/java/ml/dmlc/xgboost4j/gpu/java/CudfColumnBatch.java:63]: Just to be clear, is `DeviceDmatrix` the same thing as `DeviceQuantileDMatrix` in this context?

> Is there any necessary to separate DeviceDmatrix from DMatrix?

If this achieves the goal of an interface that makes sense, then yes.

File: Jenkinsfile [hunk 16]
```diff
@@ -64,7 +64,7 @@ pipeline {
             // The build-gpu-* builds below use Ubuntu image
             'build-gpu-cuda11.0': { BuildCUDA(cuda_version: '11.0', build_rmm: true) },
             'build-gpu-rpkg': { BuildRPackageWithCUDA(cuda_version: '10.1') },
-            'build-jvm-packages-gpu-cuda10.1': { BuildJVMPackagesWithCUDA(spark_version: '3.0.0', cuda_version: '10.1') },
+            'build-jvm-packages-gpu-cuda10.1': { BuildJVMPackagesWithCUDA(spark_version: '3.0.0', cuda_version: '11.0') },
```
Comment (hcho3) [inline Jenkinsfile:67]: Is CUDA 11 actually needed?

**Comment by trivialfis** [comment]
https://github.com/wbo4958/xgboost/pull/5

File: jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java [hunk 17]
```diff
@@ -28,7 +30,7 @@
    * This function is be called by native code during iteration and can be made as private
    * method.  We keep it as public simply to silent the linter.
    */
-  public final String getArrayInterfaceJson() {
+  private final String getArrayInterfaceJson() {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java:33]: There's a warning about this method not being used when it's private.  If you change it to public could you please also update the comments above?

Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java:33]: where is the warning? I didn't see that

Comment (wbo4958) [reply jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/ColumnBatch.java:33]: Done.

**Comment by trivialfis** [comment]
Thanks! Will take a look again today.

**Comment by wbo4958** [comment]
@trivialfis any comments on this PR?

**Review by trivialfis — APPROVED** [review]
LGTM.  In summary, this PR adds the following aggregate types:

- `Column` is an abstract type for a single column in tabular data.
- `ColumnBatch` is an abstract type for dataframe.

- `CuDFColumn` is an implementaiton of `Column` that consume cuDF column
- `CudfColumnBatch` is an implementation of `ColumnBatch` that consumes cuDF dataframe.

- `DeviceQuantileDMatrix` is the interface for quantized data.

**Comment by trivialfis** [comment]
@RAMitchell @hcho3 Would you great if you can also have another look.

**Review by RAMitchell — APPROVED** [review]
I'm comfortable merging this.
